### PR TITLE
S3 sink

### PR DESCRIPTION
**Pull request by ArtoLord**

### CODE REVIEW

File: lzy/lzy-service/src/main/java/ai/lzy/service/config/LzyServiceConfig.java [hunk 1]
```diff
@@ -30,6 +30,8 @@ public class LzyServiceConfig {
 
     private StartupPortalConfig portal;
 
+    private String s3SinkAddress = null;  // If not set, not using s3 sink
```
Comment (imakunin) [inline lzy/lzy-service/src/main/java/ai/lzy/service/config/LzyServiceConfig.java:33]: annotate it as `@Nullable`, plz

File: lzy/lzy-service/src/main/java/ai/lzy/service/workflow/StartExecutionCompanion.java [hunk 2]
```diff
@@ -323,7 +328,31 @@ public void createKafkaTopic(KafkaAdminClient kafkaAdminClient) {
...
+
+                var resp = owner.s3SinkClient.stub().submit(KafkaS3Sink.SubmitRequest.newBuilder()
+                    .setTopicName(topicName)
+                    .setStorageConfig(LMST.StorageConfig.newBuilder()
+                        .setUri("s3://" + path)
+                        .setS3(storageConfig.getS3())
+                        .build())
+                    .build());
+
+                sinkJobId = resp.getJobId();
```
Comment (imakunin) [inline lzy/lzy-service/src/main/java/ai/lzy/service/workflow/StartExecutionCompanion.java:350]: add logs, plz
smth like `Configure S3Sink, topic: {}, bucket: {}`

File: lzy/s3-sink/pom.xml [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-slf4j-impl</artifactId>
```
Comment (imakunin) [inline lzy/s3-sink/pom.xml:99]: does it really needed?

File: lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class Job {
+    private static final Logger LOG = LogManager.getLogger(Job.class);
+    private static final int BUFFER_SIZE = 1024 * 1024 * 5; // S3 multipart chunk must be at least 5Mb
+    private final AtomicReference<State> state = new AtomicReference<>(State.Created);
```
Comment (imakunin) [inline lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java:34]: add empty line between static and non-static fields, plz

File: lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    private final KafkaS3Sink.SubmitRequest request;
+    private final AtomicBoolean completed = new AtomicBoolean(false);
+
+    private final List<CompletedPart> completedParts = new ArrayList<>();
+
+
+    private final ByteBuffer s3ChunkBuffer = ByteBuffer.allocate(BUFFER_SIZE);
+
+    @Nullable private final Consumer<String, byte[]> consumer;
```
Comment (imakunin) [inline lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java:43]: place annotation on separate lines , plz

File: lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        consumer = new KafkaConsumer<>(props);
+
+        var hostAndPort = HostAndPort.fromString(endpoint);
+
+        try {
+            storageClient = S3AsyncClient.builder()
+                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(accessToken,
+                    secretToken)))
+                .region(Region.US_WEST_1)
+                .endpointOverride(new URI("http", null, hostAndPort.getHost(), hostAndPort.getPort(), null, null, null))
```
Comment (imakunin) [inline lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java:96]: why `http`? should be `https`

File: lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+        final CreateMultipartUploadResponse resp;
+        try {
+            resp = storageClient.createMultipartUpload(CreateMultipartUploadRequest.builder()
+                    .bucket(bucket)
+                    .key(key)
+                    .contentType("text/plain")
+                    .build())
+                .get();
+        } catch (InterruptedException | ExecutionException e) {
```
Comment (imakunin) [inline lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java:117]: we should retry all 5xx errors

File: lzy/s3-sink/src/main/java/ai/lzy/kafka/JobExecutor.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                    } catch (Exception e) {
+                        LOG.error("Exception while waiting for job from queue, exception is ignored: ", e);
+                    }
+                }
+            });
+        }
+    }
+
+    public String submit(Job job) {
+        var id = UUID.randomUUID().toString();
```
Comment (imakunin) [inline lzy/s3-sink/src/main/java/ai/lzy/kafka/JobExecutor.java:43]: you may wish to use `IdGenerator` from `util-common`

File: lzy/s3-sink/src/main/resources/log4j2.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+  Loggers:
+    Root:
+      level: info
+      AppenderRef:
+        - ref: Console
+        - ref: FluentBitFile
+
+    Logger:
+      - name: "ai.lzy.service.graph"
```
Comment (imakunin) [inline lzy/s3-sink/src/main/resources/log4j2.yaml:37]: it looks strange :)

File: lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java (hunk 7 above)
Comment (ArtoLord) [reply lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java:121]: S3Client will automaticly retry errors from s3

File: lzy/s3-sink/pom.xml (hunk 3 above)
Comment (ArtoLord) [reply lzy/s3-sink/pom.xml:104]: Yes, kafka client is using slf4j

File: lzy/s3-sink/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,13 @@
+FROM amazoncorretto:17
+
+COPY target/s3-sink.jar app/app.jar
+
+ENTRYPOINT ["java", \
+            "-Xmx512m", \
```
Comment (imakunin) [inline lzy/s3-sink/Dockerfile:6]: let's increase it to 1Gb

**Review by imakunin — APPROVED** [review]

**Pull request merged**